### ISSUE TITLE
firestore: Listening to snapshots fail silently when ca-certificates are missing and hangs forever

### ISSUE THREAD

**Issue by renannprado** [googleapis/google-cloud-go#1234]
## Client

Firestore

## Describe Your Environment

Docker `debian:9.6`

## Expected Behavior

Firestore snapshots should be received

## Actual Behavior

Firestore snapshots are not received, but I get no errors

Relevant code:

```go
snapshot := client.Collection("colectionName").Snapshots(ctx)
snapshot.Next()
```

By enabling the `GRPC_GO_LOG_SEVERITY_LEVEL=INFO` variable (thanks to GoogleCloudPlatform/google-cloud-go#791) I clearly see that the error is related to the certificates.
But my *guess* is that this package could report it at least, that's why I'm creating this issue.

Seems to be related to this almost-year-old PR grpc/grpc-go#1742

I think this is actually expected from the library (not to work) if the certificates are missing, but I believe they should be reported by the library instead of hanging forever (I know the bug actually lies in the gRPC, but maybe this point was not covered in the library and maybe it's possible to report it or at very least my intention is to document this behavior as an issue just in case more people goes through this).

**Workaround:**

The workaround is to install package `apt-get install -y ca-certificates` in the docker image.

**Comment by jadekler** [comment]
I thought that behavior had been "fixed"... Hm. Are you using the latest version of grpc-go?

**Comment by jadekler** [comment]
Also, are you using `grpc.WithBlock`?

**Comment by renannprado** [comment]
@jadekler I'm using `v1.16.0`

> Also, are you manually supplying grpc.WithBlock?
No

This is how I'm creating the client:

```go
	app, err := firebase.NewApp(ctx, nil, option.WithCredentialsFile(cfg.FirebaseCredentials))
	if err != nil {
		log.WithError(err).Fatal("error initializing app")
	}

	client, err := app.Firestore(ctx)

	if err != nil {
		log.WithError(err).Fatal("error while creating firestore app")
	}
```

**Comment by jadekler** [comment]
Ah! What is the `firebase` package? Perhaps they're supplying `WithBlock`..

**Comment by renannprado** [comment]
it is `cloud.google.com/go/firestore`
and the version is `0.33.1`

**Comment by jadekler** [comment]
Sorry, my question is about the fire*base* package :) e.g. where you use `app, err := firebase.NewApp`. It seems like you're getting your fire*store* client from the fire*base* package, so I'm wondering whether that package is injecting grpc options.

**Comment by renannprado** [comment]
I don't think so.
I'm using `v3.4.0` of `firebase.google.com/go`

**Comment by jadekler** [comment]
Thanks, this is quite odd. Thanks for the report, will investigate further when I find a bit of time.

**Comment by renannprado** [comment]
@jadekler I believe this is also related to GH-1232. Because if I had a way to know that the struct is in an error state, then I wouldn't attempt using it and hence the nilpointer wouldn't happen.
That's indeed odd, but my first thought is that the API would have to change to accommodate proper handling of such cases.

**Comment by jadekler** [comment]
I was able to reproduce. Filed https://github.com/grpc/grpc-go/issues/2472 to investigate.

**Comment by jadekler** [comment]
For posterity, what's happening here is that grpc returns codes.Unavailable, which the client library retries. Unfortunately there's no distinction between transient unavailability and permanent unavailability. In this case it's the latter.

**Comment by sduskis** [comment]
@jadekler, it sounds like you exposed a bug.  Can you please flip the type label to "bug"?

**Comment by jadekler** [comment]
@sduskis Not sure I'd call this a bug: everything is WAI. As mentioned before, it's a symptom of the fact that grpc's codes.Unavailable has no way to distinguish between transient and permanent unavailability. We retry codes.Unavailable because 99% of the time it's transient. However, in the case of missing ca-certificates, it's permanent.

The answer may be simple: our docs specifically call out that ca-certificates are necessary, and generally it's encouraged to set a context deadline. Both of these address the issue.

Having written all that out, I think this issue should be closed. If anyone has further thoughts on the matter please let me know!

**Comment by broady** [comment]
Having a certificate mismatch or failure of cert verification *is* a (relatively) permanent failure*, and we should be able to fail fast with a good message.

Is this error returned by gRPC? If so, we should not retry it.

* at least until the CA certs are added to the client host or TLS cert on the remote changes.

**Comment by jadekler** [comment]
> Having a certificate mismatch or failure of cert verification is a (relatively) permanent failure

Just to make sure we're not talking past each other: that was my assessment too. :)

> Is this error returned by gRPC? If so, we should not retry it.

_An_ error is returned to grpc that has no mention of ca-certificates heh:

```
rpc error: code = Unavailable desc = there is no connection available
```

To specifically check for this status, what you'd want is for it to return,

```
transport: authentication handshake failed: x509: failed to load system roots and no roots provided
```

Unfortunately I discovered in https://github.com/grpc/grpc-go/issues/1768, this is not what happens. It's not clear to me that grpc _should_, though - it _is_ still trying to connect, so in theory in the future your RPC could work.

A different way we could handle this is for the client library itself to check for cert existence before trying to create a grpc client, but that has problems, too.

**Comment by renannprado** [comment]
> The answer may be simple: our docs specifically call out that ca-certificates are necessary, and generally it's encouraged to set a context deadline. Both of these address the issue.

Do you mean that the "listening to snapshots" is not supposed to be used in a way that it "hangs forever" until snapshots are received? Are we supposed to call the `Snapshots(context.Context)` method always with a deadline?

**Comment by jadekler** [comment]
@renannprado You bring up a good point, which is that using context can't always be the answer. I've followed up with the grpc team and we've found that a small amount of plumbing is missing to get this error to the client libraries. I've PRed that here: https://github.com/grpc/grpc-go/pull/2508.

After that PR, we should be able to check for that error and not retry. :)

Re-opening this issue to track the work.
